### PR TITLE
Fix handling of empty arrays in ufunc.reduce/accumulate

### PR DESCRIPTION
**Pull request by DanisNone**
- Added handling of zero-dim arrays in `jnp.ufunc.reduce`/`accumulate`.
- Added test cases with zero dimensions to existing `reduce`/`reduceat`/`accumulate` tests.
- Did not create separate tests since they would fully duplicate existing ones, only differing in input shapes

### CODE REVIEW

File: jax/_src/numpy/ufunc_api.py [hunk 1]
```diff
@@ -387,6 +389,8 @@ def _accumulate_via_scan(self, arr: ArrayLike, axis: int = 0,
       raise ValueError("accumulate does not allow multiple axes")
     axis = canonicalize_axis(axis, np.ndim(arr))
 
+    if 0 in arr.shape:
+      return lax.full_like(arr, 0)
```
Comment (jakevdp) [inline jax/_src/numpy/ufunc_api.py]: Should this be `return lax.full(final_shape, 0, dtype)`?

File: jax/_src/numpy/ufunc_api.py [hunk 2]
```diff
@@ -387,6 +389,8 @@ def _accumulate_via_scan(self, arr: ArrayLike, axis: int = 0,
       raise ValueError("accumulate does not allow multiple axes")
     axis = canonicalize_axis(axis, np.ndim(arr))
 
+    if 0 in arr.shape:
```
Comment (jakevdp) [inline jax/_src/numpy/ufunc_api.py]: `if arr.size == 0`

File: tests/lax_numpy_ufuncs_test.py [hunk 3]
```diff
@@ -87,6 +87,7 @@ def _jnp_ufunc_props(name):
 
 broadcast_compatible_shapes = [(), (1,), (3,), (1, 3), (4, 1), (4, 3)]
 nonscalar_shapes = [(3,), (4,), (4, 3)]
+with_zerodim_shapes = nonscalar_shapes + [(0, 3), (4, 0)]
```
Comment (jakevdp) [inline tests/lax_numpy_ufuncs_test.py]: minor nit: `zerodim` implies to me that `arr.ndim == 0`. Can we do similar to `lax_numpy_test.py` and call these `empty_shapes = [(0, 3), (4, 0)]` and then define `all_shapes = nonscalar_shapes + empty_shapes`?

**Review by jakevdp — CHANGES_REQUESTED** [review]
Thanks for looking into this!

File: tests/lax_numpy_ufuncs_test.py [hunk 4]
```diff
@@ -226,13 +229,18 @@ def test_binary_ufunc_outer(self, name, lhs_shape, rhs_shape, dtype):
...
        for axis in [None, *range(-len(shape), len(shape))]],
       dtype=jtu.dtypes.floating,
   )
   def test_frompyfunc_reduce(self, func, nin, nout, identity, shape, axis, dtype):
     if (nin, nout) != (2, 1):
       self.skipTest(f"reduce requires (nin, nout)=(2, 1); got {(nin, nout)=}")
+    if identity is None and axis is None and 0 in shape:
+      self.skipTest("reduce along zero size array for ufuncs with no identity.")
+    if identity is None and axis is not None and shape[axis] == 0:
+      self.skipTest("reduce along zero-sized axis for ufuncs with no identity.")
```
Comment (jakevdp) [inline tests/lax_numpy_ufuncs_test.py]: Rather than skipping these tests, we should use `assertRaises` to assert that they raise the expected error

File: tests/lax_numpy_ufuncs_test.py [hunk 5]
```diff
@@ -255,6 +263,8 @@ def test_binary_ufunc_reduce(self, name, shape, axis, dtype):
 
     if jnp_fun.identity is None and axis is None and len(shape) > 1:
       self.skipTest("Multiple-axis reduction over non-reorderable ufunc.")
+    if jnp_fun.identity is None and axis is not None and shape[axis] == 0:
+      self.skipTest("reduce along zero-sized axis for ufuncs with no identity.")
```
Comment (jakevdp) [inline tests/lax_numpy_ufuncs_test.py]: Same here: assert that the expected error is raised

**Review by jakevdp — COMMENTED** [review]
Sorry, I missed that you had updated this (I think github doesn't notify me if you update your branch/resolve comments, but only if you comment on the PR)

One small suggestion below.

**Comment by DanisNone** [comment]
@jakevdp Made changes. Did I use assertRaises correctly?

**Review by jakevdp — APPROVED** [review]
